### PR TITLE
Nette 3.0

### PR DESCRIPTION
**Pull request by PavelJurasek**

### CODE REVIEW

**Comment by hrach** [comment]
❤️

**Comment by PavelJurasek** [comment]
Bump. Is there anything I can do to get this merged and released? Maybe access to the repo in case it is abandoned?

**Comment by sasule** [comment]
ping @enumag 
Could you please merge?

File: .travis.yml [hunk 1]
```diff
@@ -7,36 +7,9 @@ cache:
...
-  matrix:
-    - NETTE=nette-2.4-dev
-    - NETTE=nette-2.4
-
-matrix:
-  include:
-    - php: 5.6
-      env: NETTE=nette-2.4 COMPOSER_EXTRA_ARGS="--prefer-lowest --prefer-stable"
-    - php: 7.0
-      env: NETTE=nette-2.4 COVERAGE="--coverage ./coverage.xml --coverage-src ./src" TESTER_RUNTIME="phpdbg"
```
Comment (enumag) [inline .travis.yml]: It looks like you completely removed code coverage report. Any reason for that?

File: composer.json [hunk 2]
```diff
@@ -17,31 +17,17 @@
 		"issues": "https://github.com/kdyby/autowired/issues"
 	},
 	"require": {
-		"nette/di": "^2.4.7@dev",
-		"nette/application": "^2.4@dev",
-		"nette/reflection": "^2.4.2@dev",
-		"tracy/tracy": "^2.4@dev"
+		"php": "^7.1",
+		"nette/di": "^3.0",
+		"nette/application": "3.0",
```
Comment (enumag) [inline composer.json]: Would it be possible to support both 2.4 and 3.0?

File: phpstan.neon [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	level: 7
+	paths:
+		- src
+		- tests/KdybyTests
+	autoload_files:
+		- %currentWorkingDirectory%/tests/KdybyTests/ContainerTestCase.php
+		- %currentWorkingDirectory%/tests/KdybyTests/Autowired/mocks/LoremService.php
+		- %currentWorkingDirectory%/tests/KdybyTests/Autowired/mocks/IntegrationPresenter.php
+		- %currentWorkingDirectory%/tests/KdybyTests/Autowired/mocks/DatagridComponent.php
+	reportUnmatchedIgnoredErrors: false # for some reasons overriden annotation types are ok on CI
```
Comment (enumag) [inline phpstan.neon]: This is really odd, needs more investigation I think.

File: phpstan.neon [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	bootstrap: %currentWorkingDirectory%/phpstan.bootstrap.php
+	level: 7
+	paths:
+		- src
+		- tests/KdybyTests
+	autoload_files:
+		- %currentWorkingDirectory%/tests/KdybyTests/ContainerTestCase.php
+		- %currentWorkingDirectory%/tests/KdybyTests/Autowired/mocks/LoremService.php
+		- %currentWorkingDirectory%/tests/KdybyTests/Autowired/mocks/IntegrationPresenter.php
+		- %currentWorkingDirectory%/tests/KdybyTests/Autowired/mocks/DatagridComponent.php
```
Comment (enumag) [inline phpstan.neon]: Maybe add autoload-dev to composer.json instead?

File: src/Kdyby/Autowired/AutowireProperties.php [hunk 5]
```diff
@@ -121,40 +121,31 @@ private function validateProperty(Property $property, array $ignore)
...
 		$type = $this->resolveAnnotationClass($prop, $prop->getAnnotation('var'), 'var');
 		$metadata = [
 			'value' => NULL,
 			'type' => $type,
 		];
 
-		if (($args = (array) $prop->getAnnotation('autowire')) && !empty($args['factory'])) {
+		$args = (array) $prop->getAnnotation('autowire');
+
+		if (!empty($args['factory'])) {
```
Comment (enumag) [inline src/Kdyby/Autowired/AutowireProperties.php]: Better get rid of `empty`.

File: src/Kdyby/Autowired/AutowireProperties.php [hunk 6]
```diff
@@ -194,14 +186,12 @@ private function resolveAnnotationClass(\Reflector $prop, $annotationValue, $ann
...
-			$expandedType = NULL;
-			if (method_exists('Nette\Reflection\AnnotationsParser', 'expandClassName')) {
-				$expandedType = Nette\Reflection\AnnotationsParser::expandClassName($annotationValue,
-					$prop instanceof \ReflectionProperty
-						? Nette\Reflection\Helpers::getDeclaringClass($prop)
-						: $prop->getDeclaringClass()
-				);
-			}
+
+			$expandedType = Nette\Reflection\AnnotationsParser::expandClassName($annotationValue,
```
Comment (enumag) [inline src/Kdyby/Autowired/AutowireProperties.php]:
```suggestion
			$expandedType = Nette\Reflection\AnnotationsParser::expandClassName(
				$annotationValue,
```

File: src/Kdyby/Autowired/exceptions.php [hunk 7]
```diff
@@ -48,7 +48,7 @@ public function __construct($message = "", \Reflector $reflector = NULL, $previo
 	/**
 	 * @return \Reflector
 	 */
-	public function getReflector()
+	public function getReflector(): ?\Reflector
```
Comment (enumag) [inline src/Kdyby/Autowired/exceptions.php]: This doesn't match the return annotation.

File: src/Kdyby/Autowired/exceptions.php [hunk 8]
```diff
@@ -96,13 +96,13 @@ class UnexpectedValueException extends \UnexpectedValueException implements Exce
...
 	/**
-	 * @var \Reflector
+	 * @var \Reflector|null
 	 */
 	private $reflector;
 
 
 
-	public function __construct($message = "", \Reflector $reflector = NULL, $previous = NULL)
+	public function __construct($message = "", ?\Reflector $reflector = NULL, \Throwable $previous = NULL)
```
Comment (enumag) [inline src/Kdyby/Autowired/exceptions.php]: ?\Throwable

File: src/Kdyby/Autowired/exceptions.php [hunk 9]
```diff
@@ -113,7 +113,7 @@ public function __construct($message = "", \Reflector $reflector = NULL, $previo
 	/**
 	 * @return \Reflector
 	 */
-	public function getReflector()
+	public function getReflector(): ?\Reflector
```
Comment (enumag) [inline src/Kdyby/Autowired/exceptions.php]: Doesn't match annotation

File: src/Kdyby/Autowired/exceptions.php [hunk 10]
```diff
@@ -142,7 +142,7 @@ public function __construct($message = "", \Reflector $reflector = NULL, $previo
...
```
Comment (enumag) [inline src/Kdyby/Autowired/exceptions.php]: again

File: src/Kdyby/Autowired/exceptions.php [hunk 11]
```diff
@@ -155,13 +155,13 @@ class MissingServiceException extends \Nette\DI\MissingServiceException implemen
...
 	/**
-	 * @var \Reflector
+	 * @var \Reflector|null
 	 */
 	private $reflector;
 
 
 
-	public function __construct($message = "", \Reflector $reflector = NULL, $previous = NULL)
+	public function __construct($message = "", ?\Reflector $reflector = NULL, \Throwable $previous = NULL)
```
Comment (enumag) [inline src/Kdyby/Autowired/exceptions.php]: ?\Throwable

File: src/Kdyby/Autowired/exceptions.php [hunk 12]
```diff
@@ -172,7 +172,7 @@ public function __construct($message = "", \Reflector $reflector = NULL, $previo
 	/**
 	 * @return \Reflector
 	 */
-	public function getReflector()
+	public function getReflector(): ?\Reflector
```
Comment (enumag) [inline src/Kdyby/Autowired/exceptions.php]: again

**Comment by enumag** [comment]
Note that I don't have watch setup for anything in Kdyby anymore. When you need a review or something from me you need to ping me directly.

**Comment by sasule** [comment]
@enumag I have no credits on this PR, @PavelJurasek asked what to do. I am only "a man in the middle" :-)

**Comment by enumag** [comment]
Yeah, that message was meant for the both of you.

**Comment by PavelJurasek** [comment]
Comments has been incorporated. Regarding Nette 2.4 support I'd rather see the branches split, master for Nette 3 and, 1.x for Nette 2.4.

**Comment by PavelJurasek** [comment]
Could you release 1.6 or 2.0? Are you still willing to review PRs and manage releases or should we find a way to move this package under contributte or find another other way of maintenance?

**Comment by enumag** [comment]
Yeah I was just waiting for tests to pass. 2.0.0 is now released.

As for moving under contributte, you'd have to talk to Filip about that. I can't do that kind of decisions myself. I can however give you write privileges if you're interested in maintaining this package.

**Comment by PavelJurasek** [comment]
🎉thanks for the release! 
I believe there won't be many changes from now on. Hopefully the current state of maintainership will be good enough.

**Pull request merged**